Triggers.md

```diff
@@ -0,0 +1,234 @@
+Today [Valkey keyspace Notifications](https://valkey.io/topics/notifications/) offer the ability to publish events to dedicated pub/sub channels. 
+Valkey triggers add the ability to place extending logic on such keyspace events and to localize actions on the server side reducing the need to monitor and react to events on the application side.
+We propose to extend the existing [Valkey Functions](https://valkey.io/topics/functions-intro/) with the ability to define trigger function calls on specific keyspace events.
+As they are being integrated into the Valkey Functions infrastructure, Triggers are persisted as part of the user data.
+
+## Motivation 
+
+1. ***Localized actions.*** 
+   Today different application use cases rely on [Valkey keyspace notifications](https://valkey.io/topics/notifications/) in order to manage the database. For example,       consider a case were a key is referenced in different places (like SortedSet and lists). In order to support TTL logic a cleanup needs to be made when the key is evicted to remove it from the different lists or sets. It is possible to have the application listen for eviction events and perform the cleanup, however this would require the application to rely on non-persistent subscribe connections, implement application side logic and extra network hops, which can cause the cleanup to be performed long after the key was evicted.
+   For example lets take a commonly used pattern of scheduled tasks.
+   In such cases the user usually registers tasks in a ZSET with the matching execution time as the score.
+   Usually what is being done is setting a puller job in the application side which periodically reads all items in the set which have scores smaller than the current timestamp, and issue eval/fcall commands back to Valkey with the relevant tasks to run. 
+   This pattern introduces some waste as the application needs to maintain a puller job and perform the round trip back to the application in order to execute the scheduled operations.
+   With Valkey triggers this can be achieved without the need for a puller job by the user.
+   In order to achieve that we can use:
+   - ZSET z for holding scheduled tasks
+   - Key k to manage the next task execution time
+
+   The external application code will only have to add tasks to the ZSET **z** with score matching their required execution time.
+   1. once the task is added to **z** a trigger code will be executed which will take the minimal score from the ZSET **z**, and apply the diff to the current time to the TTL of key **k**.
+   2. once key **k** has expired, a trigger will be executed which will remove and execute all the tasks from the zset **z**
+
+   Here is an example library code to schedule tasks to be executed after/every several seconds
+
+   ![image](https://user-images.githubusercontent.com/88133677/225652613-aa85beee-05f1-4c5e-abe7-1d334b0e882f.png)
```

Comment:
```suggestion
   ![image](https://user-images.githubusercontent.com/88133677/225652613-aa85beee-05f1-4c5e-abe7-1d334b0e882f.png)
```
#!lua name=timerlib
local function on_expire(keys, args)
  redis.call('setex', 'timerlib_timer','1','0')
  redis.call('incr', 'timerlib_clock')
  redis.log( redis.LOG_NOTICE, 'ticktock')
end

local function schedule_task(script, secs)
        local now = tonumber(redis.call('TIME')[1])
        local in_secs =  secs + now
        redis.call('ZADD', 'timerlib_tasks', in_secs, script)
        local timeout = redis.call('ZRANGEBYSCORE', 'timerlib_tasks', '-inf', '+inf', 'WITHSCORES', 'limit', '0', '1')[2]
        redis.call('set', 'timerlib_scheduler', timeout)
        redis.call('expireat', 'timerlib_scheduler', timeout)
end

local function on_scheduler_tick(keys, args)
        local now =  tonumber(redis.call('TIME')[1])
        local ready_tasks = redis.call('ZRANGEBYSCORE', 'timerlib_tasks', '-inf', now)
        for i, task in ipairs(ready_tasks) do
                redis.call('ZPOPMIN', 'timerlib_tasks')
                local myfunc = loadstring(task)
                myfunc()
                local every = redis.call('HGET', 'timerlib_every', task)
                if every ~= nil then
                        schedule_task(task, tonumber(every))
                end
        end
        local num_tasks = tonumber(redis.call('ZCARD', 'timerlib_tasks'))
        if num_tasks > 0 then
                local timeout = redis.call('ZRANGEBYSCORE', 'timerlib_tasks', '-inf', '+inf', 'WITHSCORES', 'limit', '0', '1')[2]
                redis.call('set', 'timerlib_scheduler', timeout)
                redis.call('expireat', 'timerlib_scheduler', timeout)
        end
end

local function every(keys, args)
        schedule_task(args[2], tonumber(args[1]))
        redis.call('HSET','timerlib_every', args[2], args[1])
        return string.format("scheduled %s to happen every %s secs.",args[2], args[1])
end

local function schedule(keys, args)
        schedule_task(args[1], tonumber(args[2]))
        return string.format("scheduled %s to happen in %s secs.",args[1], args[2])
end
redis.register_trigger('on_expire', '__keyspace@0__:timerlib_timer', on_expire)
redis.register_trigger('on_scheduler_tick', '__keyspace@0__:timerlib_scheduler', on_scheduler_tick)
redis.register_function('schedule', schedule)
redis.register_function('every', every)